### PR TITLE
test: Add E2E Tests

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Overview

Added End to End style testing for the common OT-3 opentrons-emulation use cases.

The use cases are as follows:

- OT-3 Remote Source
- OT-3 Local Source
- OT-3 with Modules

Used the [Python Docker SDK](https://docker-py.readthedocs.io/en/stable/) to create a framework to introspect into an
actual emulated Docker system. From there the confirmed containers, volumes, bind mounts, and build arguments are
created as expected.

The following general checks are performed:

- Builder containers are created when expected
- Any default bind mounts are always added
- If building from local source code, that bind mount from host system to build container exists
- If building from remote source code, that bind mount from host system to build container does not exist
- After building, all executables are present and correctly configured
  Build args are configured as expected

# Changelog

## E2E Test Suite

The `e2e` folder under `test` contains everything related to defining and running End-to-End tests.

### 1. E2E Docker API

Using the [Python Docker SDK](https://docker-py.readthedocs.io/en/stable/index.html) the local Docker daemon can be
interacted with. Through this interaction, an abstraction layer is built in the form of 2 dataclasses.

- `ModuleContainers` - Any and all Opentrons Modules emulation containers on the host system
- `OT3Containers` - Any and all OT-3 Firmware emulation containers on the host system.

Both `ModuleContainers` and `OT3Containers` are comprised of [Container](https://docker-py.readthedocs.io/en/stable/containers.html) objects defined by the Docker SDK. Accessing
attributes on each of the objects is how validation is performed against an actual emulation system running in Docker.

### 2. Test Definition

An e2e test is defined by a `SystemTestDefinition` dataclass. This class defines expected results, metadata, and the
actual validation logic for the test. 

**Attributes**

- `test_id: str` - Unique test id for e2e test. This value will be used to look up test
- `yaml_config_relative_path: str` - Path to yaml config file. Relative to `opentrons-emulation` directory
- `monorepo_builder_created: bool` - Whether `MonorepoBuilderService` is expected to be created
- `ot3_firmware_builder_created: bool` - Whether `OT3FirmwareBuilderService` is expected to be created
- `opentrons_modules_builder_created: bool` - Whether `OpentronsModulesBuilderService` is expected to be created
- `local_monorepo_mounted: bool` - Whether local source code is expected to be mounted for opentrons repo
- `local_ot3_firmware_mounted: bool` - Whether local source code is expected to be mounted for ot3-firmware repo
- `local_opentrons_modules_mounted: bool` - Whether local source code is expected to be mounted for opentrons-modules repo
- `monorepo_build_args: BuildArgConfigurations` - Expected build args for `MonorepoBuilderService`. Can be `LATEST`, `COMMIT_ID`, or `LOCAL`
- `ot3_firmware_build_args: BuildArgConfigurations` - Expected build args for `OT3FirmwareBuilderService`. Can be `LATEST`, `COMMIT_ID`, or `LOCAL`
- `opentrons_modules_build_args: BuildArgConfigurations` - Expected build args for `OpentronsModulesBuilderService`. Can be `LATEST`, `COMMIT_ID`, or `LOCAL`

**Public Methods**

- `compare(ot3_system: OT3Containers, modules: ModuleContainers, mounts: ExpectedBindMounts)` - Runs checks against 
current running emulation system. Confirms current matches expected criteria and generates results.
- `is_failure()` - Run this method after running `compare` to determine whether all criteria was met.
- `print_output()` - Run this method to generate a textual output of results of the test.

**Private Methods**

`SystemTestDefinition` contains all validation logic in various private methods of the class. These are to
be called inside the `compare` method.

### 3. Constants Definition

The E2E test suite performs an extensive amount of string matching during validation. 

In effort to stay organized and provide context into constant values, all constant values are defined inside `consts.py`
in the form of [frozen dataclasses](https://docs.python.org/3/library/dataclasses.html#frozen-instances). 
All constants should be given meaningful class and attribute names to provide context into what is being validated.

### 4. Test Results Collection

To get around pytest's "fail-fast" methodology no `assert` statements are ran inside the private validation methods
of `SystemTestDefinition`. 

Instead, `SystemTestDefinition` maintains a collection of all results inside an `E2ETestOutput` object. 
Each validation step performs its assertion using
`E2ETestOutput.append_result(assertion: bool, result: TestDescription)`. The conditional statement is passed to 
`assertion` and a `TestDescription` object is passed to result. See the private methods of `SystemTestDefinition` for 
example usages.

At the end of all validation the user can the `print_output` method on `SystemTestDefinition` to print all output collected
during the test.

### 5. Adding New Tests

New tests are added inside `test_mappings.py`. Adding a `SystemTestDefinition` instance to the `_TEST_DEFS`
list adds another e2e test to the suite.

Also provides simple CLI to get test ids and/or path to test yaml file.

## CI Changes

- Add `cache-dir` input parameter to `setup-python-only` GitHub Action. This is necessary because when running e2e
  tetsing, because `opentrons-emulation` is checked out to a path of `${{ github.workspace }}/opentrons-emulation` we
  need to be able to change the cache dir from the default.
- Added `e2e-tests.yaml` GitHub Workflow to execute E2E testing.
    - Calls `make get-e2e-test-ids` to build a matrix of e2e tests to execute.
    - Executes each test in its own runner
    - Builds Docker images, starts containers, calls `refresh-dev` to build source code, then runs e2e pytests to
      validate everything is set up correctly

### Docker Setup Changes

- In Dockerfile, pushed `SOURCE_DOWNLOAD_LOCATION` build args into env variables. This is to allow them to be accessible
  to the container at runtime and allow pytest to validate that the build args were configured correctly.

### Makefile Changes

**Top-level Makefile**

- Add `docker volume prune -f` to `COMPOSE_REMOVE_COMMAND`, to remove any volumes not attached to a container. This
  cleans up all volumes that were left behind after removing the emulation containers.
- Add `refresh-dev` and `refresh-dev-ci` commands to rebuild source code. The only difference between the 2 commands
  is `refresh-dev-ci` does not run in a `tty` session because GitHub Actions doesn't like that
- Add abstracted calls to `emulation-system/Makefile` to run e2e commands

**emulation_system/Makefile**

- Change `make test` command to only run tests inside of `tests/compose_file_creator`. This is so e2e tests are not
  accidentally ran when trying to run unit tests.
- Add `make get-e2e-test-ids` command to get a list of e2e tests that can be ran
- Add `make get-e2e-test-path` command to get the path to the configuration file under test, given a test_id
- Add `make execute-e2e-test` to run an e2e test filtered by the test id

### `emulation_system` Package Changes

- Change load_containers command to be delimited by `\n` instead of space. This makes it easier for xargs to parse.
- Set `name` parameter for top level named volumes. This is necessary because docker will automatically prepend the
  docker compose project name to volume names. This causes the volumes to be named something
  like `opentrons_emulation_monorepo-wheels` instead of just `monorepo-wheels`

# Review requests

Really want to walk through this with a few people. Was a ton of work, but I think it adds a lot of value.

# Risk assessment

Medium, doesn't actually change almost any of the core functionality inside of `emulation_system`, but adds a ton of
complexity in CI.

### CODE REVIEW

File: .github/workflows/e2e-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,114 @@
+name: Opentrons Emulation End-to-End Testing
+
+on:
+  push:
+    branches:
+      - 'main'
+      - 'release-*'
+      - '.github/workflows/e2e-tests.yaml' # Make sure to remove
```
Comment (sfoster1) [inline .github/workflows/e2e-tests.yaml:8]: IMO you should keep this, removing it makes it easy for a typo in here to fail things

File: emulation_system/tests/e2e/conftest.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+from tests.e2e.fixtures.module_containers import ModuleContainers
+from tests.e2e.fixtures.ot3_containers import OT3Containers
+from tests.e2e.utilities.consts import ExpectedMount
+from tests.e2e.utilities.helper_functions import get_container, get_containers
+
+
+@pytest.fixture
+def ot3_model_under_test(
+    testing_global_em_config: OpentronsEmulationConfiguration,
+) -> Callable:
```
Comment (sfoster1) [inline emulation_system/tests/e2e/conftest.py:27]:
```suggestion
) -> Callable[[str], OT3Containers]:
```

File: emulation_system/tests/e2e/conftest.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            modules_builder=get_container(system.opentrons_modules_builder),
+        )
+
+    return _model_under_test
+
+
+@pytest.fixture
+def modules_under_test(
+    testing_global_em_config: OpentronsEmulationConfiguration,
+) -> Callable:
```
Comment (sfoster1) [inline emulation_system/tests/e2e/conftest.py:63]:
```suggestion
) -> Callable[[str], ModuleContainers]:
```

File: emulation_system/tests/e2e/conftest.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            ),
+        )
+
+    return _model_under_test
+
+
+@pytest.fixture
+def local_mounts_under_test(
+    testing_global_em_config: OpentronsEmulationConfiguration,
+) -> Callable:
```
Comment (sfoster1) [inline emulation_system/tests/e2e/conftest.py:104]:
```suggestion
) -> Callable[[str], ExpectedBindMounts]`:
```

File: emulation_system/tests/e2e/fixtures/module_containers.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+"""Dataaclass to easily access all module containers in system."""
```
Comment (sfoster1) [inline emulation_system/tests/e2e/fixtures/module_containers.py:1]:
```suggestion
"""Dataclass to easily access all module containers in system."""
```

File: emulation_system/tests/e2e/fixtures/module_containers.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+from typing import List, Optional
+
+from docker.models.containers import Container  # type: ignore[import]
+
+
+@dataclass
+class ModuleContainers:
+    """Dataclass to access all module containers emulated system easily."""
+
+    hardware_emulation_thermocycler_modules: Optional[List[Container]]
```
Comment (sfoster1) [inline emulation_system/tests/e2e/fixtures/module_containers.py:13]: Do we really need the `Optional` here? Can an empty list serve the purpose, or is there a difference in meaning between "empty list" and "list not specified"?

File: emulation_system/tests/e2e/fixtures/module_containers.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    firmware_emulation_heater_shaker_modules: Optional[List[Container]]
+    firmware_emulation_magnetic_modules: Optional[List[Container]]
+    firmware_emulation_temperature_modules: Optional[List[Container]]
+
+    @property
+    def all_modules(self) -> List[Container]:
+        """Returns list of all modules in emulated system."""
+        result_list = []
+
+        for mod_list in [
```
Comment (sfoster1) [inline emulation_system/tests/e2e/fixtures/module_containers.py:25]: not using `Optional` would simplify this

File: emulation_system/tests/e2e/fixtures/ot3_containers.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
+"""Module containg helper class to access OT-3 Containers for local system"""
```
Comment (sfoster1) [inline emulation_system/tests/e2e/fixtures/ot3_containers.py:1]:
```suggestion
"""Module containing helper class to access OT-3 Containers for local system"""
```

File: emulation_system/tests/e2e/fixtures/ot3_containers.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    gripper: Container
+    pipettes: Container
+    bootloader: Container
+    state_manager: Container
+    robot_server: Container
+    can_server: Container
+    emulator_proxy: Container
+
+    # firmware and monorepo will actually always exist
+    # but want to check that with an assert and not a type error
```
Comment (sfoster1) [inline emulation_system/tests/e2e/fixtures/ot3_containers.py:31]: mmm IMO we really want to express that some other way, then. The types here should describe the state of the initialized dataclass, after all.

One thing we could do here, if we want to have the checking deferred, is separate "stuff we wanted to build a dataclass with" and "the dataclass we built from that stuff"; have a method that gets the second from the first; and have that method raise validation errors. The "arguments" dataclass captures whether the monorepo was specified or not; but the "values" dataclass lets dependent code use it without having to check every time.

File: emulation_system/tests/e2e/test_ot3_systems.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    """
+    ot3_system: OT3Containers = ot3_model_under_test(
+        relative_path=test_def.yaml_config_relative_path
+    )
+    modules: ModuleContainers = modules_under_test(test_def.yaml_config_relative_path)
+    mounts: ExpectedBindMounts = local_mounts_under_test(
+        test_def.yaml_config_relative_path
+    )
+    test_def.compare(ot3_system, modules, mounts)
+    assert not test_def.is_failure(), test_def.print_output()
```
Comment (sfoster1) [inline emulation_system/tests/e2e/test_ot3_systems.py:32]: probably what this should be is
```python
if test_def.is_failure():
    pytest.fail(test_def.print_output(), pytrace=False)
```
(from [docs](https://docs.pytest.org/en/7.2.x/reference/reference.html#pytest-fail))
since just the assert will get a redundant and incorrect traceback.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
First off, this is awesome - great structure, love all the stuff to prevent the fast-fail, great work.

I think that we can remove some of the pytest workarounds with just a little more work.
Fundamentally, the thing we want to do is
- For each type of configuration
   -  check everything about it every time

That reads to me like everything we check being a separate pytest test - all tests run every time, regardless of whether other tests fail.

Now that's a lot of tests, and we don't really want to write them by hand, and we do want additional tests to be pretty easy to add. The right way to do this is probably with a [pytest hook](https://docs.pytest.org/en/7.2.x/reference/reference.html#std-hook-pytest_generate_tests), but that's pretty annoying, so I think the best thing might be to add yet another level of parametrization: you're already building the system test definitions in the fixtures, so have them emit tests for each thing they want to check.

**Comment by sfoster1** [comment]
> Now that's a lot of tests, and we don't really want to write them by hand, and we do want additional tests to be pretty easy to add. The right way to do this is probably with a [pytest hook](https://docs.pytest.org/en/7.2.x/reference/reference.html#std-hook-pytest_generate_tests), but that's pretty annoying, so I think the best thing might be to add yet another level of parametrization: you're already building the system test definitions in the fixtures, so have them emit tests for each thing they want to check.

Actually, a better way to do this might be to marshal everything involved into dicts and assert the dicts deep equality. That will give you everything you want: parallel non-blocking comparisons and nice diff output.

File: .github/workflows/e2e-tests.yaml (hunk 1 above)
Comment (DerekMaggio) [reply .github/workflows/e2e-tests.yaml:8]: So I totally screwed up here. 
1. `.github/workflows/e2e-tests.yaml` is for sure not a branch
2. If I add it to `push.paths` it will only run when it is one of those branches AND `.github/workflows/e2e-tests.yaml` has been modified. Github Action Docs for this [here](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-workflow-only-when-a-push-affects-specific-files). This is not what I want. I want it to run on push to those branches only. 

The fix is to remove it from `branches` and leave it as a path in `pull_request`

File: .github/workflows/yaml-sub-sanity-check.yaml [hunk 11]
```diff
@@ -49,10 +56,28 @@ jobs:
...
-              ["otie", "exposed-port", "5000"]
+              ["exposed-port", "5000", "otie"],
+              ["monorepo-source", "test"]
             ]
           input-file: ${{ github.workspace }}/samples/ot3/ot3_remote.yaml
           output-file-location: ${{ github.workspace }}/output.yaml
 
-      - name: See file
-        run: cat ${{ github.workspace }}/output.yaml
+      - name: Get first sub value
```
Comment (y3rsh) [inline .github/workflows/yaml-sub-sanity-check.yaml:65]: What is a sub value?

File: emulation_system/Makefile [hunk 12]
```diff
@@ -19,7 +19,23 @@ teardown:
...
+.PHONY: get-e2e-test-path
+get-e2e-test-path:
+	$(if $(test_id),,$(error test_id variable required))
+	@poetry run python tests/e2e/test_mappings.py get-test-path ${test_id}
+
+
+.PHONY: execute-e2e-test
+execute-e2e-test:
+	$(if $(test_id),,$(error test_id variable required))
+	poetry run pytest tests/e2e -k "${test_id}"
```
Comment (y3rsh) [inline emulation_system/Makefile:37]:
```suggestion
	@poetry run pytest tests/e2e -k "${test_id}"
```

**Review by y3rsh — APPROVED** [review]
Got a little glassy eyed about half way through.
I think we should add a make target that allows you to see what tests would run using  the `--collect-only` argument for pytest.  The output of that is a little hard to parse with parameterization but better than nothing.
